### PR TITLE
[Merged by Bors] - feat: The lexicographic sum of two locally finite orders is locally finite

### PR DESCRIPTION
**Pull request by YaelDillies**
Forward-ports https://github.com/leanprover-community/mathlib/pull/11352

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Sum/Interval.lean [hunk 1]
```diff
@@ -79,7 +77,7 @@ theorem sumLift₂_eq_empty :
     sumLift₂ f g a b = ∅ ↔
       (∀ a₁ b₁, a = inl a₁ → b = inl b₁ → f a₁ b₁ = ∅) ∧
         ∀ a₂ b₂, a = inr a₂ → b = inr b₂ → g a₂ b₂ = ∅ := by
-  refine' ⟨fun h ↦ _, fun h ↦ _⟩
+  refine' ⟨λ h ↦ _, λ h ↦ _⟩
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Interval.lean:80]: I don't think `λ` is considered idiomatic any more

File: Mathlib/Data/Sum/Interval.lean [hunk 2]
```diff
@@ -107,6 +105,116 @@ theorem sumLift₂_mono (h₁ : ∀ a b, f₁ a b ⊆ g₁ a b) (h₂ : ∀ a b,
 
 end SumLift₂
 
+section SumLexLift
+variable (f₁ f₁' : α₁ → β₁ → Finset γ₁) (f₂ f₂' : α₂ → β₂ → Finset γ₂)
+  (g₁ g₁' : α₁ → β₂ → Finset γ₁) (g₂ g₂' : α₁ → β₂ → Finset γ₂)
+
+/-- Lifts maps `α₁ → β₁ → finset γ₁`, `α₂ → β₂ → finset γ₂`, `α₁ → β₂ → finset γ₁`,
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Interval.lean:112]:
```suggestion
/-- Lifts maps `α₁ → β₁ → finset γ₁`, `α₂ → β₂ → Finset γ₂`, `α₁ → β₂ → finset γ₁`,
```

If you run `fix-comments.py` it should catch all of these.

File: Mathlib/Data/Sum/Interval.lean [hunk 3]
```diff
@@ -107,6 +105,116 @@ theorem sumLift₂_mono (h₁ : ∀ a b, f₁ a b ⊆ g₁ a b) (h₂ : ∀ a b,
...
+
+theorem sumLexLift_nonempty :
+    (sumLexLift f₁ f₂ g₁ g₂ a b).Nonempty ↔
+      (∃ a₁ b₁, a = inl a₁ ∧ b = inl b₁ ∧ (f₁ a₁ b₁).Nonempty) ∨
+        (∃ a₁ b₂, a = inl a₁ ∧ b = inr b₂ ∧ ((g₁ a₁ b₂).Nonempty ∨ (g₂ a₁ b₂).Nonempty)) ∨
+          ∃ a₂ b₂, a = inr a₂ ∧ b = inr b₂ ∧ (f₂ a₂ b₂).Nonempty :=
+  by simp [nonempty_iff_ne_empty, sum_lex_lift_eq_empty, not_and_or]
+#align finset.sum_lex_lift_nonempty Finset.sumLexLift_nonempty
+
+end SumLexLift
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Interval.lean:217]: nit: missing blank line

File: Mathlib/Data/Sum/Interval.lean [hunk 4]
```diff
@@ -209,4 +317,110 @@ theorem Ioo_inr_inr : Ioo (inr b₁ : Sum α β) (inr b₂) = (Ioo b₁ b₂).ma
...
 
+/-! ### Lexicographical sum of orders -/
+
+namespace Lex
+variable [Preorder α] [Preorder β] [OrderTop α] [OrderBot β] [LocallyFiniteOrder α]
+  [LocallyFiniteOrder β]
+
+/-- Throwaway tactic. -/
+private unsafe def simp_lex : tactic Unit :=
+`[refine to_lex.surjective.forall₃.2 _, rintro (a | a) (b | b) (c | c); simp only
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Interval.lean:328]:
```suggestion
  `[refine to_lex.surjective.forall₃.2 _, rintro (a | a) (b | b) (c | c); simp only
```

Comment (eric-wieser) [reply Mathlib/Data/Sum/Interval.lean:328]: Though I suspect you'll need `?_` not `_`

File: Mathlib/Data/Sum/Interval.lean [hunk 5]
```diff
@@ -209,4 +317,110 @@ theorem Ioo_inr_inr : Ioo (inr b₁ : Sum α β) (inr b₂) = (Ioo b₁ b₂).ma
...
+    Ioc (inrₗ b₁ : α ⊕ₗ β) (inrₗ b₂) = (Ioc b₁ b₂).map (Embedding.inr.trans toLex.toEmbedding) := by
+  rw [← Finset.map_map]; rfl
+#align sum.lex.Ioc_inr_inr Sum.Lex.Ioc_inr_inr
+
+theorem Ioo_inr_inr :
+    Ioo (inrₗ b₁ : α ⊕ₗ β) (inrₗ b₂) = (Ioo b₁ b₂).map (Embedding.inr.trans toLex.toEmbedding) := by
+  rw [← Finset.map_map]; rfl
+#align sum.lex.Ioo_inr_inr Sum.Lex.Ioo_inr_inr
+
+end Lex
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Interval.lean:425]: nit: blank line

**Review by eric-wieser — APPROVED** [review]
LGTM modulo CI issues

File: Mathlib/Data/Sum/Interval.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Data/Sum/Interval.lean:217]: Not to my standard

File: Mathlib/Data/Sum/Interval.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Data/Sum/Interval.lean]: Something is wrong with the syntax...

File: Mathlib/Data/Sum/Interval.lean [hunk 6]
```diff
@@ -3,19 +3,17 @@ Copyright (c) 2022 Yaël Dillies. All rights reserved.
...
 
 /-!
 # Finite intervals in a disjoint union
 
-This file provides the `LocallyFiniteOrder` instance for the disjoint sum of two orders.
-
-## TODO
-
-Do the same for the lexicographic sum of orders.
+This file provides the `locally_finite_order` instance for the disjoint sum and linear sum of two
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Interval.lean:15]:
```suggestion
This file provides the `LocallyFiniteOrder` instance for the disjoint sum and linear sum of two
```

File: Mathlib/Data/Sum/Interval.lean [hunk 7]
```diff
@@ -107,6 +105,116 @@ theorem sumLift₂_mono (h₁ : ∀ a b, f₁ a b ⊆ g₁ a b) (h₂ : ∀ a b,
...
+  · rfl
+  · exact map_eq_empty.2 (h.2.2 _ _ rfl rfl)
+#align finset.sum_lex_lift_eq_empty Finset.sumLexLift_eq_empty
+
+theorem sumLexLift_nonempty :
+    (sumLexLift f₁ f₂ g₁ g₂ a b).Nonempty ↔
+      (∃ a₁ b₁, a = inl a₁ ∧ b = inl b₁ ∧ (f₁ a₁ b₁).Nonempty) ∨
+        (∃ a₁ b₂, a = inl a₁ ∧ b = inr b₂ ∧ ((g₁ a₁ b₂).Nonempty ∨ (g₂ a₁ b₂).Nonempty)) ∨
+          ∃ a₂ b₂, a = inr a₂ ∧ b = inr b₂ ∧ (f₂ a₂ b₂).Nonempty :=
+  by simp [nonempty_iff_ne_empty, sumLexLift_eq_empty, not_and_or]
```
Comment (eric-wieser) [inline Mathlib/Data/Sum/Interval.lean:214]: I was able to prove this with
```
by
  simp only [nonempty_iff_ne_empty, sumLexLift_eq_empty, not_and_or, Ne.def, not_forall]
  simp
  aesop
```
but that's pretty ugly

**Comment by eric-wieser** [comment]
Can you fix up the style here?

**Review by eric-wieser — APPROVED** [review]
bors merge

**Pull request closed without merging**